iree/turbine/kernel/wave/memory_analysis/minimize_shared_allocs.py

```diff
@@ -0,0 +1,177 @@
+# Copyright 2025 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from typing import Optional
+from iree.turbine.kernel._support.tracing import CapturedTrace
+import torch.fx as fx
+import numpy as np
+
+from ...ops.wave_ops import get_custom, Allocate, SharedMemoryBarrier, NestedRegionOp
+from ..utils.graph_utils import get_users
+from ..utils.symbol_utils import subs_idxc
+from dataclasses import dataclass
+from ...lang.global_symbols import *
+from ..._support.dtype import i8
+from .solver import (
+    determine_allocations_offsets,
+)
+import math
+from ..utils.graph_utils import is_barrier_between
+
+
+def update_sort_keys(
+    trace: CapturedTrace, graph: fx.Graph, prefix: Optional[tuple] = ()
+):
+    """
+    Update the sort keys of the graph so that
+    consecutive nodes have consecutive sort keys.
+    Also, broadcast the sort keys for ops in nested graphs.
+    After this pass, the sort keys are unique and monotonically increasing.
+    For example, if we have a graph with nodes [a, b, c, d], and c is a nested
+    region with ops [e, f, g], then the sort keys will be:
+
+    a: (0,)
+    b: (1,)
+    c: (2,)
+        e: (2, 0)
+        f: (2, 1)
+        g: (2, 2)
+    d: (3,)
+
+    so that we can always say that a < b < c < e < f < g < d.
+    """
+    for i, node in enumerate(graph.nodes):
+        node._sort_key = prefix + (i,)
+        custom = get_custom(node)
+        if isinstance(custom, NestedRegionOp):
+            update_sort_keys(
+                trace,
+                trace.region_graph.subgraphs[custom.subgraph_name],
+                node._sort_key,
+            )
+
+
+@dataclass
+class LiveInterval:
+    start: tuple[int] = (10000,)
+    end: tuple[int] = (-1,)
+
+
+def compute_live_intervals(allocs: list[fx.Node]):
+    """
+    Compute the live intervals for the allocs.
+    """
+    live_intervals = {}
+    for alloc in allocs:
+        live_intervals[alloc] = LiveInterval()
+        users, _ = get_users(alloc, None)
+        for user in users:
+            if user._sort_key < live_intervals[alloc].start:
+                live_intervals[alloc].start = user._sort_key
+            if user._sort_key > live_intervals[alloc].end:
+                live_intervals[alloc].end = user._sort_key
```

Comment:
Doesn't the `_sort_key` run into the nested graph issue we were talking about?

Reply:
No because we renumber all the nodes first, see update_sort_keys above.